### PR TITLE
Add page-up page-down to editor component

### PR DESCRIPTION
**Pull request by palmtreefrb**
## Description

Added keyboard hooks, Page Up and Page Down to allow paging through the editor component.

### Type of Change
Feature request.

Only keep lines below that describe this change, then delete the rest.

- New feature (non-breaking change which adds functionality)
- UI change

## Screenshots

Peek 2021-09-16 16-22

## How Has This Been Tested?

Please describe the manual tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Press Page Down to move to the next record
- [x] Press Page Up to move to the previous record
- [x] Press Page Down to move to the next record after reaching end
- [x] Press Page Up to move to the previous record after reaching end

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have reviewed the title/description of this PR which will be used as the squashed PR commit message
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by palmtreefrb** [comment]
Screenshot from 2021-09-15 17-47-46

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (longrunningprocess) [inline src/angular-app/languageforge/lexicon/editor/editor.component.ts:110]: I'm wondering whether `onInit` is the right lifecycle hook to set this up or not?  If that's convention, then ok, however I would probably be more inclined to set the listener up after the data is back and we actually realize we have a need for the ability to move around.  Additionally, will this clean-up the listeners after the user leaves this page?

**Review by longrunningprocess — COMMENTED** [review]
Some of the template fields need to be updated before merging.

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (palmtreefrb) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:110]: In Angularjs onInit will occur after all elements have been constructed and bound.

Comment (megahirt) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:110]: Per our discussion, binding to `window` is acceptable.  However, we need to unregister the event listener on the $onDestroy() lifecycle hook.

Comment (megahirt) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:110]: There are console errors that need to be addressed before taking a look at this again.

The console errors have to do with calling the method in a context that it doesn't expect.

I will approve with the current approach of $onInit / $onDestroy

However, an alternative approach could be to use the $on('$locationChangeStart') with an eye to registering the handler when the location is changed to /editor/entry and unregister the handler when the location is changed to something else.

**Review by megahirt — CHANGES_REQUESTED** [review]
There are some console errors that need fixing, per our standup discussion

**Comment by megahirt** [comment]
@palmtreefrb please also edit the PR template, specifically filling out the summary description, type of change and then deleting the screenshots section since there aren't any

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (longrunningprocess) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:110]: > However, an alternative approach could be to use the $on('$locationChangeStart') with an eye to registering the handler when the location is changed to /editor/entry and unregister the handler when the location is changed to something else.

💡  I get it now...this is a trip wire for the editor where pg-up/pg-dn are relevant.  It's different than relying on the presence/absence of a particular DOM element.  If I'm understanding this properly, this does sound perfect!

**Comment by palmtreefrb** [comment]
Narrowed the scope when loading the listener using viewContentLoaded and added unload listener and scope destroy to the existing $window.onbeforeunload method.

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (megahirt) [inline src/angular-app/languageforge/lexicon/editor/editor.component.ts:153]: Unfortunately `$window.onbeforeunload()` is what is called when a tab or windows is closed.  So this code doesn't do anything helpful before the event handler will be destroyed as the window/tab is closed anyway.

`this.$scope.$destroy();` seems unnecessary to me.  AngularJS should be calling this automatically and we shouldn't need to explicitly call it.

Keeping with the $onInit() lifecycle hook, we should use the complementary `$onDestroy()` hook to clean up the event listener.  So the `element.unbind()` should be done inside the `$onDestoy()`.

**Review by megahirt — DISMISSED** [review]

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (palmtreefrb) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:153]: I did look at $onDestroy() and it does not get called on just leaving and navigating to another page. I thought we discussed not having the key handler listening after leaving the editor page. this.$scope.$destroy(); was just an assurance everything was cleaned up. I can remove it.

Comment (longrunningprocess) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts]: @palmtreefrb I thought Chris' idea about the locationChange was pretty good, did you find it not to be a good option?

Comment (palmtreefrb) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts]: Not a good option because it doesn't get called on page load. You have to click on a navigation button to execute the event.

Comment (longrunningprocess) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts]: ok, yeah that makes sense.  Thank you.

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (longrunningprocess) [inline src/angular-app/languageforge/lexicon/editor/editor.component.ts:151]: I still don't think this will fire if someone is just navigating around...unless this event is handled specially by Angular, it would only apply for a full page refresh, not a location change in a SPA.

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
unless my reading is wrong, this will not give you what you actually want.

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (palmtreefrb) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:151]: Pull it down and add a brakepoint there. It does actually give the behavior I was intending.

Comment (longrunningprocess) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts:151]: ok, good enough for me.  I'll approve.  Thanks.

**Review by longrunningprocess — APPROVED** [review]

**Comment by longrunningprocess** [comment]
@palmtreefrb unless you want Robin and/or Chris as well, you can remove them and merge.

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (megahirt) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts]: that's too bad that `$locationChangeStart` doesn't get fired on first page load :(

**Review by palmtreefrb — COMMENTED** [review]
Can I merge this now? Merge is still blocked and waiting for approval.

**Comment by longrunningprocess** [comment]
since you've requested Chris to approve and the last thing he did was request changes, you either need to remove him from the reviewers or make the changes he requested (if they're still relevant)

File: src/angular-app/languageforge/lexicon/editor/editor.component.ts
Comment (palmtreefrb) [reply src/angular-app/languageforge/lexicon/editor/editor.component.ts]: > since you've requested Chris to approve and the last thing he did was request changes, you either need to remove him from the reviewers or make the changes he requested (if they're still relevant)

Not an option I can find to clear this up to complete the merge. Maybe my user level does not have these options.

**Comment by rmunn** [comment]
@longrunningprocess wrote:
> since you've requested Chris to approve and the last thing he did was request changes, you either need to remove him from the reviewers or make the changes he requested (if they're still relevant)

I don't think someone who is *not* a repo admin can remove someone else's "change requested" review.

**Comment by longrunningprocess** [comment]
ok then, based upon @megahirt 's WhatsApp stand-up message, I'm going to go ahead and get this one in.

**Pull request merged**
